{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "51a211f6-e0fe-4422-970c-e4500d2e11e8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import custom"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "359efc77-633c-43e2-b3ee-ca1231a49cb3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: STOP codon not included\n"
     ]
    }
   ],
   "source": [
    "# Import package\n",
    "import custom\n",
    "\n",
    "# Start the optimizer\n",
    "opt = custom.TissueOptimizer(\"Kidney\", n_pool=100)\n",
    "\n",
    "# Optimize the eGFP sequence\n",
    "egfp = \"MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLKFICTTGKLPVPWPTLVTTLTYGVQCFSRYPDHMKQHDFFKSAMPEGYVQERTIFFKDDGNYKTRAEVKFEGDTLVNRIELKGIDFKEDGNILGHKLEYNYNSHNVYIMADKQKNGIKVNFKIRHNIEDGSVQLADHYQQNTPIGDGPVLLPDNHYLSTQSALSKDPNEKRDHMVLLEFVTAAGITLGMDELYK\"\n",
    "opt.optimize(egfp)\n",
    "\n",
    "# Select the top 10 sequences\n",
    "best_egfp_lung = opt.select_best(by={\"MFE\":\"min\",\"MFEini\":\"max\",\"CAI\":\"max\",\"CPB\":\"max\",\"ENC\":\"min\", \"GC\":float, \"UD\": float, \"UAD\": float},homopolymers=7, top=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9004adc4-5aa8-498c-be89-5d061661ed54",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>MFE</th>\n",
       "      <th>MFEini</th>\n",
       "      <th>CAI</th>\n",
       "      <th>CPB</th>\n",
       "      <th>ENC</th>\n",
       "      <th>GC</th>\n",
       "      <th>UD</th>\n",
       "      <th>UAD</th>\n",
       "      <th>Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ATGGTGAGCAAAGGGGAGGAGCTCTTTACGGGCGTCGTGCCCATTC...</td>\n",
       "      <td>-8.372571</td>\n",
       "      <td>-6.2</td>\n",
       "      <td>0.867947</td>\n",
       "      <td>-0.020429</td>\n",
       "      <td>37.399800</td>\n",
       "      <td>0.578801</td>\n",
       "      <td>0.179916</td>\n",
       "      <td>0.421199</td>\n",
       "      <td>0.794674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>ATGGTGTCGAAGGGAGAGGAGCTGTTTACTGGGGTGGTGCCCATTC...</td>\n",
       "      <td>-7.576646</td>\n",
       "      <td>-1.1</td>\n",
       "      <td>0.856405</td>\n",
       "      <td>-0.003137</td>\n",
       "      <td>37.654086</td>\n",
       "      <td>0.576011</td>\n",
       "      <td>0.178522</td>\n",
       "      <td>0.423989</td>\n",
       "      <td>0.785147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTTACAGGCGTGGTCCCCATCC...</td>\n",
       "      <td>-7.183229</td>\n",
       "      <td>-8.9</td>\n",
       "      <td>0.867920</td>\n",
       "      <td>-0.022484</td>\n",
       "      <td>37.153177</td>\n",
       "      <td>0.577406</td>\n",
       "      <td>0.170153</td>\n",
       "      <td>0.422594</td>\n",
       "      <td>0.734554</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>ATGGTGAGCAAGGGGGAGGAGCTCTTTACGGGGGTGGTGCCCATCC...</td>\n",
       "      <td>-7.535266</td>\n",
       "      <td>-4.6</td>\n",
       "      <td>0.820096</td>\n",
       "      <td>-0.029580</td>\n",
       "      <td>35.094734</td>\n",
       "      <td>0.580195</td>\n",
       "      <td>0.171548</td>\n",
       "      <td>0.419805</td>\n",
       "      <td>0.714595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>ATGGTGTCGAAGGGCGAAGAGCTGTTCACCGGCGTGGTGCCCATCC...</td>\n",
       "      <td>-7.605172</td>\n",
       "      <td>-10.4</td>\n",
       "      <td>0.870396</td>\n",
       "      <td>-0.026971</td>\n",
       "      <td>39.643786</td>\n",
       "      <td>0.570432</td>\n",
       "      <td>0.168759</td>\n",
       "      <td>0.429568</td>\n",
       "      <td>0.694625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>ATGGTGTCAAAGGGGGAGGAGCTGTTTACCGGGGTCGTGCCCATAC...</td>\n",
       "      <td>-6.932759</td>\n",
       "      <td>-4.3</td>\n",
       "      <td>0.846793</td>\n",
       "      <td>-0.008365</td>\n",
       "      <td>39.848978</td>\n",
       "      <td>0.577406</td>\n",
       "      <td>0.177127</td>\n",
       "      <td>0.422594</td>\n",
       "      <td>0.693337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>ATGGTGTCCAAGGGTGAGGAGCTGTTCACGGGGGTGGTGCCCATCC...</td>\n",
       "      <td>-8.035893</td>\n",
       "      <td>-6.9</td>\n",
       "      <td>0.858047</td>\n",
       "      <td>-0.025756</td>\n",
       "      <td>45.326195</td>\n",
       "      <td>0.573222</td>\n",
       "      <td>0.181311</td>\n",
       "      <td>0.426778</td>\n",
       "      <td>0.660801</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>ATGGTGTCGAAGGGCGAGGAGCTCTTCACAGGAGTGGTGCCCATCC...</td>\n",
       "      <td>-7.621944</td>\n",
       "      <td>-11.9</td>\n",
       "      <td>0.829009</td>\n",
       "      <td>0.001600</td>\n",
       "      <td>34.488923</td>\n",
       "      <td>0.574616</td>\n",
       "      <td>0.174338</td>\n",
       "      <td>0.425384</td>\n",
       "      <td>0.648776</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>ATGGTGAGTAAGGGGGAAGAACTGTTCACGGGCGTGGTCCCGATCC...</td>\n",
       "      <td>-7.342163</td>\n",
       "      <td>-6.7</td>\n",
       "      <td>0.861117</td>\n",
       "      <td>-0.049145</td>\n",
       "      <td>39.243037</td>\n",
       "      <td>0.566248</td>\n",
       "      <td>0.172943</td>\n",
       "      <td>0.433752</td>\n",
       "      <td>0.642987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>ATGGTGTCGAAGGGGGAGGAACTCTTCACCGGGGTGGTGCCCATCC...</td>\n",
       "      <td>-7.310188</td>\n",
       "      <td>-8.8</td>\n",
       "      <td>0.859270</td>\n",
       "      <td>-0.055087</td>\n",
       "      <td>39.489372</td>\n",
       "      <td>0.573222</td>\n",
       "      <td>0.174338</td>\n",
       "      <td>0.426778</td>\n",
       "      <td>0.642218</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             Sequence       MFE  MFEini  \\\n",
       "15  ATGGTGAGCAAAGGGGAGGAGCTCTTTACGGGCGTCGTGCCCATTC... -8.372571    -6.2   \n",
       "99  ATGGTGTCGAAGGGAGAGGAGCTGTTTACTGGGGTGGTGCCCATTC... -7.576646    -1.1   \n",
       "75  ATGGTGAGCAAGGGCGAGGAGCTGTTTACAGGCGTGGTCCCCATCC... -7.183229    -8.9   \n",
       "32  ATGGTGAGCAAGGGGGAGGAGCTCTTTACGGGGGTGGTGCCCATCC... -7.535266    -4.6   \n",
       "76  ATGGTGTCGAAGGGCGAAGAGCTGTTCACCGGCGTGGTGCCCATCC... -7.605172   -10.4   \n",
       "98  ATGGTGTCAAAGGGGGAGGAGCTGTTTACCGGGGTCGTGCCCATAC... -6.932759    -4.3   \n",
       "63  ATGGTGTCCAAGGGTGAGGAGCTGTTCACGGGGGTGGTGCCCATCC... -8.035893    -6.9   \n",
       "53  ATGGTGTCGAAGGGCGAGGAGCTCTTCACAGGAGTGGTGCCCATCC... -7.621944   -11.9   \n",
       "56  ATGGTGAGTAAGGGGGAAGAACTGTTCACGGGCGTGGTCCCGATCC... -7.342163    -6.7   \n",
       "55  ATGGTGTCGAAGGGGGAGGAACTCTTCACCGGGGTGGTGCCCATCC... -7.310188    -8.8   \n",
       "\n",
       "         CAI       CPB        ENC        GC        UD       UAD     Score  \n",
       "15  0.867947 -0.020429  37.399800  0.578801  0.179916  0.421199  0.794674  \n",
       "99  0.856405 -0.003137  37.654086  0.576011  0.178522  0.423989  0.785147  \n",
       "75  0.867920 -0.022484  37.153177  0.577406  0.170153  0.422594  0.734554  \n",
       "32  0.820096 -0.029580  35.094734  0.580195  0.171548  0.419805  0.714595  \n",
       "76  0.870396 -0.026971  39.643786  0.570432  0.168759  0.429568  0.694625  \n",
       "98  0.846793 -0.008365  39.848978  0.577406  0.177127  0.422594  0.693337  \n",
       "63  0.858047 -0.025756  45.326195  0.573222  0.181311  0.426778  0.660801  \n",
       "53  0.829009  0.001600  34.488923  0.574616  0.174338  0.425384  0.648776  \n",
       "56  0.861117 -0.049145  39.243037  0.566248  0.172943  0.433752  0.642987  \n",
       "55  0.859270 -0.055087  39.489372  0.573222  0.174338  0.426778  0.642218  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "best_egfp_lung"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "85208563-b4ce-487f-ad99-7aae9dbdec39",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>MFE</th>\n",
       "      <th>MFEini</th>\n",
       "      <th>CAI</th>\n",
       "      <th>CPB</th>\n",
       "      <th>ENC</th>\n",
       "      <th>Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>ATGGTGAGCAAGGGTGAGGAGCTGTTTACTGGTGTGGTGCCGATCC...</td>\n",
       "      <td>-7.788401</td>\n",
       "      <td>-6.2</td>\n",
       "      <td>0.825462</td>\n",
       "      <td>-0.006591</td>\n",
       "      <td>35.230725</td>\n",
       "      <td>0.695552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ATGGTGTCCAAGGGGGAGGAGCTGTTTACTGGGGTGGTGCCAATCC...</td>\n",
       "      <td>-7.156113</td>\n",
       "      <td>-3.9</td>\n",
       "      <td>0.869224</td>\n",
       "      <td>-0.054007</td>\n",
       "      <td>40.304473</td>\n",
       "      <td>0.673032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>ATGGTCTCCAAGGGAGAGGAGCTGTTTACTGGGGTGGTGCCGATCC...</td>\n",
       "      <td>-7.521787</td>\n",
       "      <td>-6.3</td>\n",
       "      <td>0.858592</td>\n",
       "      <td>-0.017169</td>\n",
       "      <td>41.406881</td>\n",
       "      <td>0.671707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ATGGTGTCCAAGGGCGAGGAGCTGTTTACGGGTGTGGTCCCTATCC...</td>\n",
       "      <td>-7.572100</td>\n",
       "      <td>-7.0</td>\n",
       "      <td>0.828129</td>\n",
       "      <td>0.026304</td>\n",
       "      <td>40.284963</td>\n",
       "      <td>0.663456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>ATGGTGTCCAAGGGAGAGGAGCTCTTTACGGGGGTGGTGCCCATCC...</td>\n",
       "      <td>-7.212382</td>\n",
       "      <td>-7.0</td>\n",
       "      <td>0.876334</td>\n",
       "      <td>-0.031641</td>\n",
       "      <td>41.382824</td>\n",
       "      <td>0.650946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>ATGGTGTCCAAAGGGGAGGAGCTGTTTACGGGCGTCGTCCCCATCC...</td>\n",
       "      <td>-7.441223</td>\n",
       "      <td>-9.4</td>\n",
       "      <td>0.822960</td>\n",
       "      <td>0.034225</td>\n",
       "      <td>35.198568</td>\n",
       "      <td>0.650788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>ATGGTCTCCAAGGGCGAGGAGCTGTTTACTGGCGTGGTGCCAATAC...</td>\n",
       "      <td>-7.257524</td>\n",
       "      <td>-6.5</td>\n",
       "      <td>0.884197</td>\n",
       "      <td>-0.041717</td>\n",
       "      <td>43.959737</td>\n",
       "      <td>0.644286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>ATGGTGTCCAAGGGCGAGGAGCTGTTTACCGGTGTGGTCCCCATAC...</td>\n",
       "      <td>-8.347649</td>\n",
       "      <td>-7.5</td>\n",
       "      <td>0.848331</td>\n",
       "      <td>-0.061149</td>\n",
       "      <td>40.516153</td>\n",
       "      <td>0.642719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ATGGTGTCCAAAGGGGAGGAGCTGTTCACGGGAGTCGTGCCCATCC...</td>\n",
       "      <td>-7.592320</td>\n",
       "      <td>-6.8</td>\n",
       "      <td>0.857878</td>\n",
       "      <td>-0.052615</td>\n",
       "      <td>38.698983</td>\n",
       "      <td>0.642655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTTACGGGGGTGGTGCCAATCC...</td>\n",
       "      <td>-7.530408</td>\n",
       "      <td>-7.6</td>\n",
       "      <td>0.860668</td>\n",
       "      <td>-0.015654</td>\n",
       "      <td>42.536369</td>\n",
       "      <td>0.641621</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             Sequence       MFE  MFEini  \\\n",
       "50  ATGGTGAGCAAGGGTGAGGAGCTGTTTACTGGTGTGGTGCCGATCC... -7.788401    -6.2   \n",
       "27  ATGGTGTCCAAGGGGGAGGAGCTGTTTACTGGGGTGGTGCCAATCC... -7.156113    -3.9   \n",
       "59  ATGGTCTCCAAGGGAGAGGAGCTGTTTACTGGGGTGGTGCCGATCC... -7.521787    -6.3   \n",
       "1   ATGGTGTCCAAGGGCGAGGAGCTGTTTACGGGTGTGGTCCCTATCC... -7.572100    -7.0   \n",
       "79  ATGGTGTCCAAGGGAGAGGAGCTCTTTACGGGGGTGGTGCCCATCC... -7.212382    -7.0   \n",
       "47  ATGGTGTCCAAAGGGGAGGAGCTGTTTACGGGCGTCGTCCCCATCC... -7.441223    -9.4   \n",
       "75  ATGGTCTCCAAGGGCGAGGAGCTGTTTACTGGCGTGGTGCCAATAC... -7.257524    -6.5   \n",
       "84  ATGGTGTCCAAGGGCGAGGAGCTGTTTACCGGTGTGGTCCCCATAC... -8.347649    -7.5   \n",
       "28  ATGGTGTCCAAAGGGGAGGAGCTGTTCACGGGAGTCGTGCCCATCC... -7.592320    -6.8   \n",
       "37  ATGGTGAGCAAGGGCGAGGAGCTGTTTACGGGGGTGGTGCCAATCC... -7.530408    -7.6   \n",
       "\n",
       "         CAI       CPB        ENC     Score  \n",
       "50  0.825462 -0.006591  35.230725  0.695552  \n",
       "27  0.869224 -0.054007  40.304473  0.673032  \n",
       "59  0.858592 -0.017169  41.406881  0.671707  \n",
       "1   0.828129  0.026304  40.284963  0.663456  \n",
       "79  0.876334 -0.031641  41.382824  0.650946  \n",
       "47  0.822960  0.034225  35.198568  0.650788  \n",
       "75  0.884197 -0.041717  43.959737  0.644286  \n",
       "84  0.848331 -0.061149  40.516153  0.642719  \n",
       "28  0.857878 -0.052615  38.698983  0.642655  \n",
       "37  0.860668 -0.015654  42.536369  0.641621  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "best_egfp_lung"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4f7f10bf-c062-454e-86ae-05254badfa5b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pkg_resources\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import RNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "38e1a85e-3de1-4649-a44e-349a6bfefd05",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Load data required for optimization\n",
    "GENETIC_CODE = {\n",
    "    'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n",
    "    'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "    'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n",
    "    'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',\n",
    "    'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
    "    'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "    'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "    'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "    'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n",
    "    'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "    'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n",
    "    'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "    'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
    "    'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
    "    'TAC':'Y', 'TAT':'Y', 'TAA':'_', 'TAG':'_',\n",
    "    'TGC':'C', 'TGT':'C', 'TGA':'_', 'TGG':'W'}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1068f7eb-2669-48e8-ad7b-4e966423af6a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def load_data(file):\n",
    "    '''\n",
    "    Return a dataframe of the required file.\n",
    "\n",
    "    '''\n",
    "    # stream = pkg_resources.resource_stream(__name__, file)\n",
    "    return pd.read_csv(file, index_col=0)\n",
    "\n",
    "\n",
    "# def load_data(file):\n",
    "#     '''\n",
    "#     Return a dataframe of the required file.\n",
    "\n",
    "#     '''\n",
    "#     stream = pkg_resources.resource_stream(__name__, file)\n",
    "#     return pd.read_csv(stream, index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "3cfffa45-ee63-4cb9-92b6-ed8b145234c9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "codon_weights = load_data('data/CUSTOM_codon_weights.csv')\n",
    "codon_ratios = load_data('data/CUSTOM_tissue_ratios.csv')\n",
    "codon_freq = load_data('data/CUSTOM_codonfreq_CoCoPuts.csv')\n",
    "codpair_freq = load_data('data/CUSTOM_codonpairsfreq_CoCoPuts.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "230c54c6-d489-4f43-8325-091b0dee6e2b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(36, 61)\n",
      "(36, 61)\n",
      "(64, 1)\n",
      "(4096, 1)\n"
     ]
    }
   ],
   "source": [
    "print(codon_weights.shape)\n",
    "print(codon_ratios.shape)\n",
    "print(codon_freq.shape)\n",
    "print(codpair_freq.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "902fb10e-529c-46ab-937f-e44a95505539",
   "metadata": {},
   "source": [
    "# Separate functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "5a512bc6-ed50-4a24-a397-d98285983115",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def check_is_optimized(optimizer):\n",
    "    '''\n",
    "    Raises an error if the sequence is not optimized.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    None.\n",
    "\n",
    "    '''\n",
    "    if not hasattr(optimizer,\"pool\"):\n",
    "        raise TypeError(\"This TissueOptimizer instance is not optimized yet. \"\n",
    "                        \"Call 'optimize' with appropriate arguments before \"\n",
    "                        \"using this optimizer.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "75dee986-b2cb-44f8-b7d3-431c6f7c0650",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def relative_codons():\n",
    "    '''\n",
    "    Computes normalized codon abundances for each amino-acid family.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    codnorm: DataFrame\n",
    "\n",
    "    '''\n",
    "    if \"codnorm\" not in globals():\n",
    "        global codnorm\n",
    "        codnorm = pd.DataFrame()\n",
    "        AAs = list(set(GENETIC_CODE.values())); AAs.remove(\"_\")\n",
    "        for species in codon_freq.columns:\n",
    "            for aa in AAs:\n",
    "                cod_aa = [c for c in GENETIC_CODE.keys() if GENETIC_CODE[c]==aa]\n",
    "                max_aa = codon_freq.loc[cod_aa,species].max()\n",
    "                for cod in cod_aa:\n",
    "                    codnorm.loc[cod,species] = codon_freq.loc[cod,species]/max_aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "f06c9700-019c-449c-bbe0-e8b9706b47f1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Homo_sapiens</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GTA</th>\n",
       "      <td>0.299282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTC</th>\n",
       "      <td>0.521935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTG</th>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTT</th>\n",
       "      <td>0.458421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATA</th>\n",
       "      <td>0.436740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGT</th>\n",
       "      <td>0.715199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCA</th>\n",
       "      <td>0.721550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCC</th>\n",
       "      <td>0.879171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCG</th>\n",
       "      <td>0.206216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCT</th>\n",
       "      <td>0.862660</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Homo_sapiens\n",
       "GTA      0.299282\n",
       "GTC      0.521935\n",
       "GTG      1.000000\n",
       "GTT      0.458421\n",
       "ATA      0.436740\n",
       "..            ...\n",
       "AGT      0.715199\n",
       "TCA      0.721550\n",
       "TCC      0.879171\n",
       "TCG      0.206216\n",
       "TCT      0.862660\n",
       "\n",
       "[61 rows x 1 columns]"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codnorm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "cba4878d-2c9e-47df-b485-72c5de688c91",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def compute_CPS():\n",
    "    '''\n",
    "    Computes the Codon Pair Scores based on the observed vs expected\n",
    "    frequency. The CPS is + for overrepresented pairs and - for \n",
    "    underrepresented pairs. The expected frequency is computed to be \n",
    "    independent both of amino acid frequency and of codon bias.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    CPSs: DataFrame\n",
    "\n",
    "    '''\n",
    "    if \"CPSs\" not in globals():\n",
    "        global CPSs\n",
    "        CPSs = pd.DataFrame()\n",
    "        for species in codpair_freq.columns:\n",
    "            for pair in codpair_freq.index:\n",
    "                if (pair[0:3] not in [\"TAA\",\"TGA\",\"TAG\"]) and (pair[3:6] not in [\"TAA\",\"TGA\",\"TAG\"]):\n",
    "                    obs_pair = codpair_freq.loc[pair,species]\n",
    "                    # Compute codon occurrences\n",
    "                    cod_A = codon_freq.loc[pair[0:3],species]\n",
    "                    cod_B = codon_freq.loc[pair[3:6],species]\n",
    "                    # Compute AA occurrences\n",
    "                    aaA_cods = [c for c in GENETIC_CODE.keys() if GENETIC_CODE[c]==GENETIC_CODE[pair[0:3]]]\n",
    "                    aaB_cods = [c for c in GENETIC_CODE.keys() if GENETIC_CODE[c]==GENETIC_CODE[pair[3:6]]]\n",
    "                    aa_A = codon_freq.loc[aaA_cods,species].sum()\n",
    "                    aa_B = codon_freq.loc[aaB_cods,species].sum()\n",
    "                    # Compute AA pair occurences\n",
    "                    aaAB_codpairs = [c1+c2 for c1 in aaA_cods for c2 in aaB_cods]\n",
    "                    aa_AB = codpair_freq.loc[aaAB_codpairs,species].sum()\n",
    "                    # Compute expected frequency\n",
    "                    exp_pair = ((cod_A*cod_B)/(aa_A*aa_B))*aa_AB\n",
    "                    # Compute CPS\n",
    "                    CPSs.loc[pair,species] = np.log(obs_pair/exp_pair)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "c2890a31-fe71-4df2-9259-c833242a1038",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Homo_sapiens</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TTTTTT</th>\n",
       "      <td>-0.349047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTTTC</th>\n",
       "      <td>-0.404618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTTTA</th>\n",
       "      <td>0.117131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTTTG</th>\n",
       "      <td>-0.088855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTCTT</th>\n",
       "      <td>0.083706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGGAGG</th>\n",
       "      <td>-0.004143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGGGGT</th>\n",
       "      <td>-0.080464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGGGGC</th>\n",
       "      <td>0.311483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGGGGA</th>\n",
       "      <td>-0.471404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGGGGG</th>\n",
       "      <td>-0.790970</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3721 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Homo_sapiens\n",
       "TTTTTT     -0.349047\n",
       "TTTTTC     -0.404618\n",
       "TTTTTA      0.117131\n",
       "TTTTTG     -0.088855\n",
       "TTTCTT      0.083706\n",
       "...              ...\n",
       "GGGAGG     -0.004143\n",
       "GGGGGT     -0.080464\n",
       "GGGGGC      0.311483\n",
       "GGGGGA     -0.471404\n",
       "GGGGGG     -0.790970\n",
       "\n",
       "[3721 rows x 1 columns]"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CPSs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "7fa2a97b-fbfc-4b4d-9791-8e3adca49452",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def action(metric,to_do):\n",
    "    '''\n",
    "    Takes a list of computed values and converts them to minimize or maximize\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    metric : ndarray\n",
    "        Values to optimize\n",
    "    \n",
    "    to_do : {\"min\",\"max\"}\n",
    "        Min or Max\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    norm : ndarray\n",
    "\n",
    "    '''\n",
    "    if to_do==\"max\":\n",
    "        norm = metric\n",
    "        return norm\n",
    "    elif to_do==\"min\":\n",
    "        norm = -metric\n",
    "        return norm\n",
    "    else:\n",
    "        raise TypeError(\"Invalid 'by' argument. Please specify either 'max' \"\n",
    "                        \"or 'min' to indicate whether each \"\n",
    "                        \"metric should be maximized or minimized.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "bb098a55-1d53-42dc-8489-c195ec3e7c03",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "class TissueOptimizer:\n",
    "    '''\n",
    "    Optimize object, which contains all required methods for tissue-optimizing \n",
    "    the codons an amino-acid or nucleotide sequence. Codons are selected based \n",
    "    on the Random Forest features that define tissue-specificities, which\n",
    "    are considered as the probability of optimizing each codon. Directionality\n",
    "    is based on tissue-specific SDA ratios. Among the pool of generated \n",
    "    sequences, the best ones are selected based several commonly used metrics\n",
    "    (CAI, ENC, CPB, MFE, etc.).\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    tissue: {\"Lung\",\"Breast\",\"Skin\",\"Spleen\",\"Heart\",\"Liver\",\"Salivarygland\",\n",
    "             \"Muscle...Skeletal\",\"Tonsil\",\"Smallintestine\",\"Placenta\",\n",
    "             \"Appendices\",\"Testis\",\"Rectum\",\"Urinarybladder\",\"Prostate\",\n",
    "             \"Esophagus\",\"Kidney\",\"Thyroid\",\"Lymphnode\",\"Artery\",\"Brain\",\n",
    "             \"Nerve...Tibial\",\"Gallbladder\",\"Uterus\",\"Pituitary\",\"Colon\",\n",
    "             \"Vagina\",\"Duodenum\",\"Fat\",\"Stomach\",\"Adrenal\",\"Fallopiantube\",\n",
    "             \"Smoothmuscle\",\"Pancreas\",\"Ovary\"}\n",
    "        Tissue to which optimize the sequence\n",
    "    \n",
    "    n_pool: int, default=100\n",
    "        The number of sequences in the generated pool of optimized sequences.\n",
    "        \n",
    "    degree: float between 0-1, default=0.5\n",
    "        Percentage of codons to optimize. Higher values lead to optimizing \n",
    "        all codons. Lower values optimize only the most clearly \n",
    "        tissue-specific codons.\n",
    "    \n",
    "    prob_original: float between 0-1, default=0.0\n",
    "        Extent to which original codons are conserved. Higher values lead to \n",
    "        more conservative optimizations.\n",
    "    \n",
    "    Attributes\n",
    "    ----------\n",
    "    pool : list\n",
    "        Minimum Free Energy of optimized sequences. Predicted secondary \n",
    "        structures are based on the Vienna RNA Package (Lorenz et al., 2011).\n",
    "    \n",
    "    codonprob : dict\n",
    "        For each amino acid, it contains a nested dictionary matching each\n",
    "        codon with (1) its tissue-specific weight and (2) its directionality\n",
    "        (True if its inclusion is favored in that tissue, False otherwise).\n",
    "    \n",
    "    sequence : str\n",
    "        Original sequence to be optimized.\n",
    "    \n",
    "    Examples\n",
    "    --------\n",
    "    >>> import custom\n",
    "    >>> opt = TissueOptimizer(\"kidney\", n_pool=50)\n",
    "    >>> seq = \"MVSKGEELFTGVVPILVELDGDVNGHKFSVSG\"\n",
    "    >>> opt.optimize(seq)\n",
    "    >>> best_seq = opt.select_best(top=10)\n",
    "    \n",
    "    '''\n",
    "    def __init__(self, tissue, n_pool=100, degree=0.5, prob_original=0.0):\n",
    "        if tissue in codon_weights.index:\n",
    "            self.tissue = tissue\n",
    "            self.n_pool = n_pool\n",
    "            self.degree = degree\n",
    "            self.prob_original = prob_original\n",
    "            \n",
    "            # Get directionality and probabilities for tissue optimization\n",
    "            # \"degree\" determines to which extent are unclear codons optimized\n",
    "            threshold = np.percentile(codon_ratios.loc[self.tissue,:].abs(),(1.0-self.degree)*100)\n",
    "            codonprob = {}\n",
    "            for aa in set(GENETIC_CODE.values()):\n",
    "                aa_codons = [c for c in GENETIC_CODE.keys() if np.logical_and((GENETIC_CODE[c] is aa),(c not in [\"ATG\",\"TGG\",\"TAA\",\"TGA\",\"TAG\"]))]\n",
    "                # Avoid deterministic behaviour of codons with weight=0\n",
    "                if any(codon_weights.loc[self.tissue,aa_codons]):\n",
    "                    weight_bkg = codon_weights.loc[self.tissue,aa_codons].mean()\n",
    "                    codweights = {c:codon_weights.loc[self.tissue,c]+weight_bkg for c in aa_codons}\n",
    "                else:\n",
    "                    codweights = {c:codon_weights.loc[self.tissue,c] for c in aa_codons}\n",
    "                sumweights = sum(codweights.values())\n",
    "                if sumweights==0:\n",
    "                    sumweights = 1.0\n",
    "                coddirection = {c:codon_ratios.loc[self.tissue,c]>0 if np.abs(codon_ratios.loc[self.tissue,c])>threshold else np.nan for c in aa_codons}\n",
    "                codonprob[aa] = {c:[codweights[c]/sumweights,coddirection[c]] for c in aa_codons}\n",
    "            self.codonprob = codonprob\n",
    "        else:\n",
    "            raise TypeError(\"Invalid 'tissue' argument. Allowed tissues are: \"\n",
    "                            \"Lung, Breast, Skin, Spleen, Heart, Liver, Salivarygland, \"\n",
    "                            \"Muscle...Skeletal, Tonsil, Smallintestine, Placenta, \"\n",
    "                            \"Appendices, Testis, Rectum, Urinarybladder, Prostate, \"\n",
    "                            \"Esophagus, Kidney, Thyroid, Lymphnode, Artery, Brain, \"\n",
    "                            \"Nerve...Tibial, Gallbladder, Uterus, Pituitary, Colon, \"\n",
    "                            \"Vagina, Duodenum, Fat, Stomach, Adrenal, Fallopiantube, \"\n",
    "                            \"Smoothmuscle, Pancreas, Ovary.\")\n",
    "    \n",
    "    def optimize(self, sequence):\n",
    "        '''\n",
    "        Create a pool of optimized sequences\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        sequence : str\n",
    "            Original nucleotide coding sequence to optimize without stop codon. \n",
    "            It also accepts amino acid sequences. In the latter, the 'prob_original' \n",
    "            argument is overrun as the optimization runs from scratch with no prior \n",
    "            knowledge.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        self : object\n",
    "\n",
    "        '''\n",
    "        \n",
    "        def optaa(self, aa):\n",
    "            codons = list(self.codonprob[aa].keys())\n",
    "            newcodon = \"\"\n",
    "            while not newcodon:\n",
    "                if len(codons)==1:\n",
    "                    newcodon = codons[0]\n",
    "                else:\n",
    "                    probs = np.array([self.codonprob[aa][c][0] for c in codons])\n",
    "                    probs = probs/np.sum(probs)\n",
    "                    action = np.random.choice(codons,p=probs)\n",
    "                    if self.codonprob[aa][action][1]==True:\n",
    "                        newcodon = action\n",
    "                    elif self.codonprob[aa][action][1]==False:\n",
    "                        codons.remove(action)\n",
    "                    else:\n",
    "                        if np.random.uniform()<=0.5:\n",
    "                            newcodon = action\n",
    "                        else:\n",
    "                            codons.remove(action)\n",
    "                    \n",
    "            return newcodon\n",
    "        \n",
    "        self.sequence = sequence.upper()\n",
    "        \n",
    "        # Identify type of sequence\n",
    "        if all([s in \"ATCG\" for s in self.sequence]) and len(self.sequence)%3==0: # DNA sequence\n",
    "            seqcodons = [self.sequence[n:n+3] for n in range(0,len(self.sequence),3)]\n",
    "            # Create pool of optimized sequences\n",
    "            pool = []\n",
    "            for n in range(self.n_pool):\n",
    "                newseq = []\n",
    "                unifrand = np.random.uniform(size=len(seqcodons))\n",
    "                for n,codon in enumerate(seqcodons):\n",
    "                    # prob_original allows to take a conservative optimization if wanted\n",
    "                    if codon in [\"ATG\",\"TGG\",\"TAA\",\"TGA\",\"TAG\"]: # do not touch stop codons nor ATG or TGG\n",
    "                        newseq.append(codon)\n",
    "                    elif unifrand[n]>=self.prob_original:\n",
    "                        aa = GENETIC_CODE[codon]\n",
    "                        newseq.append(optaa(self,aa))\n",
    "                    else:\n",
    "                        newseq.append(codon)\n",
    "                pool.append(\"\".join(newseq))\n",
    "\n",
    "        elif all([s in GENETIC_CODE.values() for s in self.sequence]): # AA sequence\n",
    "            # Create pool of optimized sequences\n",
    "            pool = []\n",
    "            for n in range(self.n_pool):\n",
    "                newseq = []\n",
    "                for aa in self.sequence:\n",
    "                    if aa==\"M\":\n",
    "                        newseq.append(\"ATG\")\n",
    "                    elif aa==\"W\":\n",
    "                        newseq.append(\"TGG\")\n",
    "                    else:\n",
    "                        newseq.append(optaa(self,aa))\n",
    "                pool.append(\"\".join(newseq))\n",
    "            print(\"Warning: STOP codon not included\")\n",
    "            self.pool = pool\n",
    "        else:\n",
    "            raise TypeError(\"The sequence is not valid.\")\n",
    "    \n",
    "    def MFE(self):\n",
    "        '''\n",
    "        Calculates the Minimum Free Energy of the body (excludes first 40 nt) \n",
    "        of optimized sequences.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        MFE: list\n",
    "\n",
    "        '''\n",
    "        check_is_optimized(self)\n",
    "        MFEs = []\n",
    "        for seq in self.pool:\n",
    "            # mRNA folding in vivo happens while the CDS is being transcribed.\n",
    "            # To reproduce it as much as possible, get the average of a \n",
    "            # sliding window of 40 nucleotides along the sequence.\n",
    "            windows = [seq[w:w+40] for w in range(40,(len(seq)-39))]\n",
    "            mfe_temp = []\n",
    "            for w in windows:\n",
    "                mfe_temp.append(RNA.fold(w)[1])\n",
    "            MFEs.append(np.mean(mfe_temp))\n",
    "        return MFEs\n",
    "    \n",
    "    def MFEini(self):\n",
    "        '''\n",
    "        Calculates the Minimum Free Energy of the first 40 nt of optimized \n",
    "        sequences.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        MFE: list\n",
    "\n",
    "        '''\n",
    "        check_is_optimized(self)\n",
    "        MFEs = []\n",
    "        for seq in self.pool:\n",
    "            MFEs.append(RNA.fold(seq[:40])[1])\n",
    "        return MFEs\n",
    "    \n",
    "    def CAI(self):\n",
    "        '''\n",
    "        Calculate the Codon Adaptation Index, based on the similarity to the \n",
    "        codon composition of the human genome as a whole (source: CoCoPuts).\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        CAI: list\n",
    "\n",
    "        '''\n",
    "        check_is_optimized(self)\n",
    "        relative_codons() # Compute RCU of human genome\n",
    "        CAIs = []\n",
    "        for seq in self.pool:\n",
    "            seqcodons = [seq[n:n+3] for n in range(0,len(seq),3)]\n",
    "            codus = {c:seqcodons.count(c) for c in set(seqcodons) if c not in [\"TAA\",\"TGA\",\"TAG\"]}\n",
    "            cai_codus = [(codnorm.loc[c,\"Homo_sapiens\"]**(1/sum(codus.values())))**codus[c] for c in codus.keys()]\n",
    "            CAIs.append(np.prod(cai_codus))\n",
    "        return CAIs\n",
    "    \n",
    "    def CPB(self):\n",
    "        '''\n",
    "        Calculate the Codon Pair Bias, based on the similarity to the \n",
    "        codon pair composition of the human genome as a whole \n",
    "        (source: CoCoPuts).\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        CPB: list\n",
    "\n",
    "        '''\n",
    "        check_is_optimized(self)\n",
    "        compute_CPS()\n",
    "        CPBs = []\n",
    "        for seq in self.pool:\n",
    "            seqcodpairs = [seq[n:n+6] for n in range(0,len(seq),3) if (len(seq[n:n+6])==6)and(seq[n+3:n+6] not in [\"TAA\",\"TGA\",\"TAG\"])]\n",
    "            codscores = [CPSs.loc[pair,\"Homo_sapiens\"] for pair in seqcodpairs]\n",
    "            CPBs.append(np.mean(codscores))\n",
    "        return CPBs\n",
    "    \n",
    "    def ENC(self):\n",
    "        '''\n",
    "        Calculate the Effective Number of Codons, which is a measure of codon\n",
    "        evenness. A value of 20 means that all 100% codons are biased towards \n",
    "        the most common codon, while 61 corresponds to no bias at all.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        ENC: list\n",
    "\n",
    "        '''\n",
    "        check_is_optimized(self)\n",
    "        codon_families = {aa:list(GENETIC_CODE.values()).count(aa) for aa in GENETIC_CODE.values() if aa!=\"_\"}\n",
    "        ENCs = []\n",
    "        for seq in self.pool:\n",
    "            seqcodons = [seq[n:n+3] for n in range(0,len(seq),3)]\n",
    "            codus = {c:seqcodons.count(c) for c in set(seqcodons) if c not in [\"TAA\",\"TGA\",\"TAG\"]}\n",
    "            ENC_family = {}\n",
    "            for f in [2,3,4,6]:\n",
    "                fam_aa = [aa for aa in codon_families.keys() if codon_families[aa]==f]\n",
    "                ENCfam = []\n",
    "                for aa in fam_aa:\n",
    "                    cod_aa = [c for c in GENETIC_CODE.keys() if GENETIC_CODE[c]==aa]\n",
    "                    if any([c in codus.keys() for c in cod_aa]):\n",
    "                        codus_family = np.array([codus[c] if c in codus.keys() else 0.0 for c in cod_aa])\n",
    "                        codus_sum = codus_family.sum()\n",
    "                        if codus_sum>1:\n",
    "                            p_family = codus_family/codus_sum\n",
    "                            numerator = (codus_sum - 1.0)\n",
    "                            denominator = (codus_sum * np.sum([p**2 for p in p_family]) - 1.0)\n",
    "                            if numerator!=0 and denominator!=0:\n",
    "                                ENCfam.append(numerator/denominator)\n",
    "                ENC_family[f] = ENCfam\n",
    "            # Compute contributions of each family\n",
    "            enc_codus = 2.0 + 9.0*np.mean(ENC_family[2]) + 1.0*np.mean(ENC_family[3]) + 5.0*np.mean(ENC_family[4]) + 3.0*np.mean(ENC_family[6])\n",
    "            ENCs.append(enc_codus)\n",
    "        return ENCs\n",
    "    \n",
    "    def GC(self):\n",
    "        '''\n",
    "        Calculate the GC content of sequences.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        GC: list\n",
    "\n",
    "        '''\n",
    "        check_is_optimized(self)\n",
    "        GCs = []\n",
    "        for seq in self.pool:\n",
    "            GCs.append((seq.count(\"G\")+seq.count(\"C\"))/len(seq))\n",
    "        return GCs\n",
    "    \n",
    "    def U_count(self):\n",
    "        '''\n",
    "        Calculate the number of Uridine in sequences.\n",
    "        \n",
    "        Suresh: Counting T and U for now, Needs revisions\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        Us: list\n",
    "\n",
    "        '''\n",
    "        check_is_optimized(self)\n",
    "        Us = []\n",
    "        for seq in self.pool:\n",
    "            Us.append((seq.count(\"T\")+seq.count(\"U\"))/len(seq))\n",
    "        return Us\n",
    "  \n",
    "    def UA_count(self):\n",
    "        '''\n",
    "        Calculate the ratio of Uracil/Adenine in sequence.\n",
    "        \n",
    "        Suresh: Counting T and U both for now, Needs revisions*** \n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        UAs: list\n",
    "\n",
    "        '''\n",
    "        check_is_optimized(self)\n",
    "        UAs = []\n",
    "        for seq in self.pool:\n",
    "            UAs.append((seq.count(\"T\")+seq.count(\"U\")+seq.count(\"A\"))/len(seq))\n",
    "        return UAs\n",
    "    \n",
    "    def select_best(self, by={\"MFE\":\"min\",\"CAI\":\"max\",\"CPB\":\"max\",\"ENC\":\"min\"},\n",
    "                    homopolymers=0, exclude_motifs = [], top=None):\n",
    "        '''\n",
    "        Sort the pool of generated sequences based on different metrics, and\n",
    "        output the best ones.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        by : {\"MFE\":\"min/max\",\"MFEini\":\"min/max\",\"CAI\":\"min/max\",\"CPB\":\"min/max\",\n",
    "              \"ENC\":\"min/max\",\"GC\":float}, default={\"MFE\":\"min\",\"CAI\":\"max\",\"CPB\":\"max\",\"ENC\":\"min\"}\n",
    "            Metrics to use in the evaluation of sequences and whether to\n",
    "            maximize or minimize. For GC, user specified target float\n",
    "            \n",
    "        homopolymers : int, default=0\n",
    "            Removes sequences containing homopolymer sequences of n repeats.\n",
    "            \n",
    "        exclude_motifs : list, default=[]\n",
    "            Removes sequences including certain motifs. Excessively short \n",
    "            motifs are recomended against, since the probability of they \n",
    "            appearing by chance is high.\n",
    "            \n",
    "        top : int or None, default=None\n",
    "            Output only the top N sequences. If None, all sorted sequences are\n",
    "            outputted.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        best = DataFrame\n",
    "        '''\n",
    "        \n",
    "        check_is_optimized(self)\n",
    "        select_df = pd.DataFrame(index = range(self.n_pool))\n",
    "        select_df[\"Sequence\"] = self.pool\n",
    "        norm_df = pd.DataFrame(index = range(self.n_pool))\n",
    "        for c in by:\n",
    "            if c==\"MFE\":\n",
    "                metric = np.array(self.MFE())\n",
    "                norm = action(metric,by[c])\n",
    "            elif c==\"MFEini\":\n",
    "                metric = np.array(self.MFEini())\n",
    "                norm = action(metric,by[c])\n",
    "            elif c==\"CAI\":\n",
    "                metric = np.array(self.CAI())\n",
    "                norm = action(metric,by[c])\n",
    "            elif c==\"CPB\":\n",
    "                metric = np.array(self.CPB())\n",
    "                norm = action(metric,by[c])\n",
    "            elif c==\"ENC\":\n",
    "                metric = np.array(self.ENC())\n",
    "                norm = action(metric,by[c])\n",
    "            elif c==\"GC\":\n",
    "                metric = np.array(self.GC())\n",
    "                # norm = action(np.abs(metric-by[c]),\"min\")\n",
    "                norm = action(np.abs(metric),\"max\")\n",
    "            elif c==\"UD\":\n",
    "                metric = np.array(self.U_count())\n",
    "                norm = action(np.abs(metric),\"min\")  \n",
    "            elif c==\"UAD\":\n",
    "                metric = np.array(self.UA_count())\n",
    "                norm = action(np.abs(metric),\"min\")  \n",
    "            else:\n",
    "                raise TypeError(\"Invalid 'by' argument.\")\n",
    "            \n",
    "            select_df[c] = metric\n",
    "            \n",
    "            # Normalize between 0 and 1\n",
    "            norm_metric = (norm-np.min(norm))/np.ptp(norm)\n",
    "            norm_df[c] = norm_metric\n",
    "        # Create a score based on metrics\n",
    "        select_df[\"Score\"] = norm_df.mean(axis=1)\n",
    "        # Filter out homopolymers\n",
    "        if homopolymers>0:\n",
    "            homo = [\"\".join([nt]*homopolymers) for nt in [\"A\",\"T\",\"C\",\"G\"]]\n",
    "            for h in homo:\n",
    "                idx = np.array([h in seq for seq in select_df.Sequence])\n",
    "                select_df = select_df.loc[np.logical_not(idx),:]\n",
    "        # Filter out motifs\n",
    "        for m in exclude_motifs:\n",
    "            idx = np.array([m in seq for seq in select_df.Sequence])\n",
    "            select_df = select_df.loc[np.logical_not(idx),:]\n",
    "        # Sort and select best candidates\n",
    "        best = select_df.sort_values(by=\"Score\", ascending=False)\n",
    "        if top:\n",
    "            best = best.head(n=top)\n",
    "        return best"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "610e15be-008b-4e5a-8186-db9b58622adb",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: STOP codon not included\n"
     ]
    }
   ],
   "source": [
    "# Start the optimizer\n",
    "opt = TissueOptimizer(\"Kidney\", n_pool=100)\n",
    "\n",
    "# Optimize the eGFP sequence\n",
    "egfp = \"MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLKFICTTGKLPVPWPTLVTTLTYGVQCFSRYPDHMKQHDFFKSAMPEGYVQERTIFFKDDGNYKTRAEVKFEGDTLVNRIELKGIDFKEDGNILGHKLEYNYNSHNVYIMADKQKNGIKVNFKIRHNIEDGSVQLADHYQQNTPIGDGPVLLPDNHYLSTQSALSKDPNEKRDHMVLLEFVTAAGITLGMDELYK\"\n",
    "egfp = \"MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLKFICTTGKLPVPWPTLVTTLTYGVQCFSRYPDHMKQHDFFKSAMPEGYVQEITLGMDELYK\"\n",
    "\n",
    "opt.optimize(egfp)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "5ba8eb35-3868-4158-961f-02a00f800488",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating MFE\n",
      "Calculating MFEini\n",
      "Calculating CAI\n",
      "Calculating CPB\n",
      "Calculating ENC\n",
      "Calculating GC\n",
      "Calculating UD\n",
      "[0.18553459119496854, 0.18867924528301888, 0.19811320754716982, 0.19811320754716982, 0.1918238993710692, 0.1949685534591195, 0.1918238993710692, 0.1949685534591195, 0.18238993710691823, 0.19811320754716982, 0.1949685534591195, 0.20125786163522014, 0.18238993710691823, 0.20125786163522014, 0.20440251572327045, 0.2138364779874214, 0.20440251572327045, 0.1761006289308176, 0.20440251572327045, 0.1918238993710692, 0.17295597484276728, 0.1949685534591195, 0.18553459119496854, 0.2138364779874214, 0.20754716981132076, 0.18553459119496854, 0.20440251572327045, 0.18238993710691823, 0.18238993710691823, 0.19811320754716982, 0.20125786163522014, 0.19811320754716982, 0.1949685534591195, 0.18553459119496854, 0.1918238993710692, 0.18238993710691823, 0.2138364779874214, 0.1949685534591195, 0.1918238993710692, 0.1792452830188679, 0.1918238993710692, 0.20125786163522014, 0.22327044025157233, 0.19811320754716982, 0.1949685534591195, 0.18553459119496854, 0.20125786163522014, 0.20125786163522014, 0.20125786163522014, 0.19811320754716982, 0.1918238993710692, 0.18867924528301888, 0.22012578616352202, 0.18867924528301888, 0.1949685534591195, 0.18238993710691823, 0.21069182389937108, 0.18553459119496854, 0.18867924528301888, 0.18553459119496854, 0.20440251572327045, 0.20440251572327045, 0.1792452830188679, 0.18553459119496854, 0.18238993710691823, 0.1949685534591195, 0.18553459119496854, 0.20125786163522014, 0.2138364779874214, 0.1949685534591195, 0.1918238993710692, 0.20125786163522014, 0.1918238993710692, 0.18238993710691823, 0.20125786163522014, 0.2138364779874214, 0.18553459119496854, 0.1949685534591195, 0.18238993710691823, 0.20125786163522014, 0.18867924528301888, 0.21069182389937108, 0.20754716981132076, 0.20125786163522014, 0.2169811320754717, 0.1918238993710692, 0.18867924528301888, 0.20440251572327045, 0.1949685534591195, 0.20125786163522014, 0.1949685534591195, 0.20125786163522014, 0.1918238993710692, 0.21069182389937108, 0.18553459119496854, 0.1918238993710692, 0.19811320754716982, 0.1949685534591195, 0.18867924528301888, 0.20125786163522014]\n",
      "[0.18553459 0.18867925 0.19811321 0.19811321 0.1918239  0.19496855\n",
      " 0.1918239  0.19496855 0.18238994 0.19811321 0.19496855 0.20125786\n",
      " 0.18238994 0.20125786 0.20440252 0.21383648 0.20440252 0.17610063\n",
      " 0.20440252 0.1918239  0.17295597 0.19496855 0.18553459 0.21383648\n",
      " 0.20754717 0.18553459 0.20440252 0.18238994 0.18238994 0.19811321\n",
      " 0.20125786 0.19811321 0.19496855 0.18553459 0.1918239  0.18238994\n",
      " 0.21383648 0.19496855 0.1918239  0.17924528 0.1918239  0.20125786\n",
      " 0.22327044 0.19811321 0.19496855 0.18553459 0.20125786 0.20125786\n",
      " 0.20125786 0.19811321 0.1918239  0.18867925 0.22012579 0.18867925\n",
      " 0.19496855 0.18238994 0.21069182 0.18553459 0.18867925 0.18553459\n",
      " 0.20440252 0.20440252 0.17924528 0.18553459 0.18238994 0.19496855\n",
      " 0.18553459 0.20125786 0.21383648 0.19496855 0.1918239  0.20125786\n",
      " 0.1918239  0.18238994 0.20125786 0.21383648 0.18553459 0.19496855\n",
      " 0.18238994 0.20125786 0.18867925 0.21069182 0.20754717 0.20125786\n",
      " 0.21698113 0.1918239  0.18867925 0.20440252 0.19496855 0.20125786\n",
      " 0.19496855 0.20125786 0.1918239  0.21069182 0.18553459 0.1918239\n",
      " 0.19811321 0.19496855 0.18867925 0.20125786]\n",
      "Calculating UAD\n",
      "[0.41194969 0.41823899 0.43081761 0.41823899 0.41823899 0.42138365\n",
      " 0.40880503 0.42138365 0.40251572 0.41823899 0.42138365 0.41823899\n",
      " 0.39937107 0.43396226 0.42767296 0.43081761 0.41509434 0.40880503\n",
      " 0.4245283  0.42138365 0.38679245 0.42138365 0.39937107 0.4245283\n",
      " 0.41194969 0.42138365 0.4245283  0.40566038 0.41194969 0.41509434\n",
      " 0.41823899 0.41194969 0.41823899 0.41194969 0.42767296 0.40251572\n",
      " 0.43396226 0.42138365 0.40880503 0.39308176 0.41194969 0.41194969\n",
      " 0.45283019 0.41194969 0.41509434 0.41509434 0.4245283  0.41509434\n",
      " 0.41823899 0.43396226 0.40251572 0.41509434 0.44339623 0.39622642\n",
      " 0.40880503 0.40251572 0.43710692 0.41194969 0.40880503 0.40880503\n",
      " 0.4245283  0.4245283  0.38993711 0.40566038 0.40880503 0.42138365\n",
      " 0.39622642 0.43081761 0.4245283  0.41509434 0.40880503 0.43081761\n",
      " 0.42138365 0.39308176 0.43081761 0.43396226 0.39622642 0.41509434\n",
      " 0.41194969 0.42138365 0.41194969 0.42138365 0.42767296 0.41194969\n",
      " 0.44025157 0.41509434 0.39308176 0.43710692 0.39308176 0.4245283\n",
      " 0.43396226 0.4245283  0.44654088 0.43396226 0.40566038 0.4245283\n",
      " 0.41509434 0.41823899 0.41194969 0.41509434]\n"
     ]
    }
   ],
   "source": [
    "# Select the top 10 sequences\n",
    "best_egfp_lung = opt.select_best(by={\"MFE\":\"min\",\"MFEini\":\"max\",\"CAI\":\"max\",\"CPB\":\"max\",\"ENC\":\"min\", \"GC\":float, \"UD\": float, \"UAD\": float},homopolymers=7, top=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "11583bd9-a81a-4e92-af01-ee7636bdcf6e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>MFE</th>\n",
       "      <th>MFEini</th>\n",
       "      <th>CAI</th>\n",
       "      <th>CPB</th>\n",
       "      <th>ENC</th>\n",
       "      <th>GC</th>\n",
       "      <th>UD</th>\n",
       "      <th>UAD</th>\n",
       "      <th>Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>ATGGTGTCCAAGGGAGAGGAGCTGTTTACAGGCGTGGTGCCCATCC...</td>\n",
       "      <td>-9.285356</td>\n",
       "      <td>-3.8</td>\n",
       "      <td>0.833304</td>\n",
       "      <td>0.018563</td>\n",
       "      <td>38.725845</td>\n",
       "      <td>0.606918</td>\n",
       "      <td>0.182390</td>\n",
       "      <td>0.393082</td>\n",
       "      <td>0.812072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>ATGGTCAGCAAGGGCGAGGAGCTGTTTACCGGGGTGGTCCCCATCC...</td>\n",
       "      <td>-9.763180</td>\n",
       "      <td>-7.2</td>\n",
       "      <td>0.833644</td>\n",
       "      <td>0.020744</td>\n",
       "      <td>47.675471</td>\n",
       "      <td>0.603774</td>\n",
       "      <td>0.185535</td>\n",
       "      <td>0.396226</td>\n",
       "      <td>0.773464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTCACGGGCGTGGTGCCCATCC...</td>\n",
       "      <td>-8.759833</td>\n",
       "      <td>-9.8</td>\n",
       "      <td>0.820354</td>\n",
       "      <td>0.015913</td>\n",
       "      <td>35.246429</td>\n",
       "      <td>0.610063</td>\n",
       "      <td>0.179245</td>\n",
       "      <td>0.389937</td>\n",
       "      <td>0.731872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>ATGGTGTCCAAGGGCGAGGAGCTGTTTACAGGGGTGGTGCCTATCC...</td>\n",
       "      <td>-9.015063</td>\n",
       "      <td>-4.7</td>\n",
       "      <td>0.843849</td>\n",
       "      <td>-0.042490</td>\n",
       "      <td>38.074016</td>\n",
       "      <td>0.603774</td>\n",
       "      <td>0.185535</td>\n",
       "      <td>0.396226</td>\n",
       "      <td>0.721047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ATGGTGAGCAAGGGGGAGGAGCTGTTCACGGGAGTGGTGCCGATCC...</td>\n",
       "      <td>-9.088284</td>\n",
       "      <td>-7.6</td>\n",
       "      <td>0.833147</td>\n",
       "      <td>-0.102954</td>\n",
       "      <td>33.725949</td>\n",
       "      <td>0.613208</td>\n",
       "      <td>0.172956</td>\n",
       "      <td>0.386792</td>\n",
       "      <td>0.703471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>ATGGTGTCCAAGGGCGAGGAGCTGTTTACGGGAGTGGTCCCCATCC...</td>\n",
       "      <td>-7.612134</td>\n",
       "      <td>-6.8</td>\n",
       "      <td>0.830695</td>\n",
       "      <td>0.005283</td>\n",
       "      <td>40.688953</td>\n",
       "      <td>0.606918</td>\n",
       "      <td>0.179245</td>\n",
       "      <td>0.393082</td>\n",
       "      <td>0.703236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>ATGGTGTCCAAAGGGGAGGAGCTGTTTACGGGCGTCGTGCCGATCC...</td>\n",
       "      <td>-9.732636</td>\n",
       "      <td>-10.0</td>\n",
       "      <td>0.841278</td>\n",
       "      <td>-0.027500</td>\n",
       "      <td>33.707542</td>\n",
       "      <td>0.606918</td>\n",
       "      <td>0.194969</td>\n",
       "      <td>0.393082</td>\n",
       "      <td>0.689692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>ATGGTCAGCAAGGGCGAGGAGCTGTTCACGGGGGTCGTGCCCATCC...</td>\n",
       "      <td>-8.344770</td>\n",
       "      <td>-6.8</td>\n",
       "      <td>0.894418</td>\n",
       "      <td>-0.049499</td>\n",
       "      <td>40.385190</td>\n",
       "      <td>0.597484</td>\n",
       "      <td>0.182390</td>\n",
       "      <td>0.402516</td>\n",
       "      <td>0.687544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ATGGTGTCAAAAGGCGAGGAGCTCTTTACGGGGGTGGTCCCAATCC...</td>\n",
       "      <td>-9.270293</td>\n",
       "      <td>-6.5</td>\n",
       "      <td>0.858145</td>\n",
       "      <td>-0.060513</td>\n",
       "      <td>42.998122</td>\n",
       "      <td>0.594340</td>\n",
       "      <td>0.182390</td>\n",
       "      <td>0.405660</td>\n",
       "      <td>0.675071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ATGGTGTCCAAAGGCGAGGAGCTGTTCACAGGGGTGGTGCCCATAC...</td>\n",
       "      <td>-7.946444</td>\n",
       "      <td>-5.5</td>\n",
       "      <td>0.893701</td>\n",
       "      <td>-0.057582</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.588050</td>\n",
       "      <td>0.182390</td>\n",
       "      <td>0.411950</td>\n",
       "      <td>0.636752</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             Sequence       MFE  MFEini  \\\n",
       "73  ATGGTGTCCAAGGGAGAGGAGCTGTTTACAGGCGTGGTGCCCATCC... -9.285356    -3.8   \n",
       "76  ATGGTCAGCAAGGGCGAGGAGCTGTTTACCGGGGTGGTCCCCATCC... -9.763180    -7.2   \n",
       "62  ATGGTGAGCAAGGGCGAGGAGCTGTTCACGGGCGTGGTGCCCATCC... -8.759833    -9.8   \n",
       "66  ATGGTGTCCAAGGGCGAGGAGCTGTTTACAGGGGTGGTGCCTATCC... -9.015063    -4.7   \n",
       "20  ATGGTGAGCAAGGGGGAGGAGCTGTTCACGGGAGTGGTGCCGATCC... -9.088284    -7.6   \n",
       "39  ATGGTGTCCAAGGGCGAGGAGCTGTTTACGGGAGTGGTCCCCATCC... -7.612134    -6.8   \n",
       "88  ATGGTGTCCAAAGGGGAGGAGCTGTTTACGGGCGTCGTGCCGATCC... -9.732636   -10.0   \n",
       "35  ATGGTCAGCAAGGGCGAGGAGCTGTTCACGGGGGTCGTGCCCATCC... -8.344770    -6.8   \n",
       "27  ATGGTGTCAAAAGGCGAGGAGCTCTTTACGGGGGTGGTCCCAATCC... -9.270293    -6.5   \n",
       "28  ATGGTGTCCAAAGGCGAGGAGCTGTTCACAGGGGTGGTGCCCATAC... -7.946444    -5.5   \n",
       "\n",
       "         CAI       CPB        ENC        GC        UD       UAD     Score  \n",
       "73  0.833304  0.018563  38.725845  0.606918  0.182390  0.393082  0.812072  \n",
       "76  0.833644  0.020744  47.675471  0.603774  0.185535  0.396226  0.773464  \n",
       "62  0.820354  0.015913  35.246429  0.610063  0.179245  0.389937  0.731872  \n",
       "66  0.843849 -0.042490  38.074016  0.603774  0.185535  0.396226  0.721047  \n",
       "20  0.833147 -0.102954  33.725949  0.613208  0.172956  0.386792  0.703471  \n",
       "39  0.830695  0.005283  40.688953  0.606918  0.179245  0.393082  0.703236  \n",
       "88  0.841278 -0.027500  33.707542  0.606918  0.194969  0.393082  0.689692  \n",
       "35  0.894418 -0.049499  40.385190  0.597484  0.182390  0.402516  0.687544  \n",
       "27  0.858145 -0.060513  42.998122  0.594340  0.182390  0.405660  0.675071  \n",
       "28  0.893701 -0.057582        NaN  0.588050  0.182390  0.411950  0.636752  "
      ]
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "best_egfp_lung"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "8cbbe541-b0f1-4981-a102-bb65b0915bfd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>MFE</th>\n",
       "      <th>MFEini</th>\n",
       "      <th>CAI</th>\n",
       "      <th>CPB</th>\n",
       "      <th>ENC</th>\n",
       "      <th>Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ATGGTGAGCAAAGGCGAAGAACTGTTTACGGGCGTGGTGCCCATCC...</td>\n",
       "      <td>-7.610815</td>\n",
       "      <td>-5.1</td>\n",
       "      <td>0.860807</td>\n",
       "      <td>-0.017548</td>\n",
       "      <td>41.163463</td>\n",
       "      <td>0.706884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ATGGTGTCCAAGGGGGAGGAACTGTTCACGGGCGTGGTGCCGATCC...</td>\n",
       "      <td>-6.990752</td>\n",
       "      <td>-7.2</td>\n",
       "      <td>0.889493</td>\n",
       "      <td>-0.026077</td>\n",
       "      <td>37.923699</td>\n",
       "      <td>0.680873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>ATGGTGTCCAAGGGCGAGGAGCTGTTCACGGGAGTGGTCCCGATAC...</td>\n",
       "      <td>-7.999373</td>\n",
       "      <td>-7.1</td>\n",
       "      <td>0.835539</td>\n",
       "      <td>-0.018418</td>\n",
       "      <td>39.378394</td>\n",
       "      <td>0.670250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>ATGGTGTCGAAAGGCGAGGAACTGTTTACAGGCGTGGTGCCAATCC...</td>\n",
       "      <td>-7.695611</td>\n",
       "      <td>-4.8</td>\n",
       "      <td>0.837233</td>\n",
       "      <td>-0.012299</td>\n",
       "      <td>44.801380</td>\n",
       "      <td>0.649849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>ATGGTGTCCAAGGGCGAGGAGCTGTTTACAGGGGTGGTGCCGATCC...</td>\n",
       "      <td>-7.464890</td>\n",
       "      <td>-4.7</td>\n",
       "      <td>0.851330</td>\n",
       "      <td>-0.031404</td>\n",
       "      <td>41.876098</td>\n",
       "      <td>0.648878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>ATGGTGAGCAAGGGTGAGGAACTCTTCACGGGCGTGGTGCCGATAC...</td>\n",
       "      <td>-7.554859</td>\n",
       "      <td>-8.7</td>\n",
       "      <td>0.842522</td>\n",
       "      <td>0.004122</td>\n",
       "      <td>40.339701</td>\n",
       "      <td>0.643531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>ATGGTGTCCAAGGGAGAGGAGCTGTTCACTGGTGTGGTGCCGATCC...</td>\n",
       "      <td>-7.144514</td>\n",
       "      <td>-6.2</td>\n",
       "      <td>0.825245</td>\n",
       "      <td>0.018274</td>\n",
       "      <td>39.536896</td>\n",
       "      <td>0.636397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTTACGGGCGTGGTGCCCATCC...</td>\n",
       "      <td>-7.408777</td>\n",
       "      <td>-7.6</td>\n",
       "      <td>0.861615</td>\n",
       "      <td>-0.029717</td>\n",
       "      <td>40.722488</td>\n",
       "      <td>0.628250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>ATGGTGTCCAAGGGCGAGGAACTGTTCACAGGTGTGGTGCCGATCC...</td>\n",
       "      <td>-7.440439</td>\n",
       "      <td>-5.6</td>\n",
       "      <td>0.861537</td>\n",
       "      <td>-0.043537</td>\n",
       "      <td>43.345629</td>\n",
       "      <td>0.621274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>ATGGTGAGCAAAGGAGAGGAGCTGTTTACGGGCGTGGTCCCAATCC...</td>\n",
       "      <td>-7.687461</td>\n",
       "      <td>-7.3</td>\n",
       "      <td>0.859158</td>\n",
       "      <td>-0.044617</td>\n",
       "      <td>43.374581</td>\n",
       "      <td>0.610592</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             Sequence       MFE  MFEini  \\\n",
       "16  ATGGTGAGCAAAGGCGAAGAACTGTTTACGGGCGTGGTGCCCATCC... -7.610815    -5.1   \n",
       "19  ATGGTGTCCAAGGGGGAGGAACTGTTCACGGGCGTGGTGCCGATCC... -6.990752    -7.2   \n",
       "91  ATGGTGTCCAAGGGCGAGGAGCTGTTCACGGGAGTGGTCCCGATAC... -7.999373    -7.1   \n",
       "81  ATGGTGTCGAAAGGCGAGGAACTGTTTACAGGCGTGGTGCCAATCC... -7.695611    -4.8   \n",
       "34  ATGGTGTCCAAGGGCGAGGAGCTGTTTACAGGGGTGGTGCCGATCC... -7.464890    -4.7   \n",
       "78  ATGGTGAGCAAGGGTGAGGAACTCTTCACGGGCGTGGTGCCGATAC... -7.554859    -8.7   \n",
       "39  ATGGTGTCCAAGGGAGAGGAGCTGTTCACTGGTGTGGTGCCGATCC... -7.144514    -6.2   \n",
       "79  ATGGTGAGCAAGGGCGAGGAGCTGTTTACGGGCGTGGTGCCCATCC... -7.408777    -7.6   \n",
       "99  ATGGTGTCCAAGGGCGAGGAACTGTTCACAGGTGTGGTGCCGATCC... -7.440439    -5.6   \n",
       "89  ATGGTGAGCAAAGGAGAGGAGCTGTTTACGGGCGTGGTCCCAATCC... -7.687461    -7.3   \n",
       "\n",
       "         CAI       CPB        ENC     Score  \n",
       "16  0.860807 -0.017548  41.163463  0.706884  \n",
       "19  0.889493 -0.026077  37.923699  0.680873  \n",
       "91  0.835539 -0.018418  39.378394  0.670250  \n",
       "81  0.837233 -0.012299  44.801380  0.649849  \n",
       "34  0.851330 -0.031404  41.876098  0.648878  \n",
       "78  0.842522  0.004122  40.339701  0.643531  \n",
       "39  0.825245  0.018274  39.536896  0.636397  \n",
       "79  0.861615 -0.029717  40.722488  0.628250  \n",
       "99  0.861537 -0.043537  43.345629  0.621274  \n",
       "89  0.859158 -0.044617  43.374581  0.610592  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "best_egfp_lung"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
